framework/test/driver/ResourceCompilerTest.cpp

```diff
@@ -422,28 +536,569 @@ static void testrcFailureModes(const std::string& rcbinpath, const std::string&
+  manifest.close();
+}
+
+// Create a fake shared library which will be used to test embedding invalid manifest data
+// using usResourceCompiler.
+void createDummyBundle(const std::string tempdir, const std::string bundle_name)
+{
+  // Create a binary file filled with random numbers
+  // to test usResourceCompiler functionality.
+  auto create_mock_dll = [&tempdir](const std::string& dllname,
+                                    const std::array<char, 5>& dat)
+  {
+    std::string dll_path(tempdir + dllname);
+    std::ofstream dll(dll_path.c_str());
+    if (!dll.is_open())
+    {
+      throw std::runtime_error("Couldn't open " + dll_path);
+    }
+    dll.write(dat.data(), dat.size()); //binary write
+    dll.close();
+  };
+  std::array<char, 5> data1 = { {2, 4, 6, 8, 10} };
+  create_mock_dll(bundle_name, data1);
+}
+
+// A helper function designed to validate whether embedding an invalid manifest
+// to a fake shared library worked or not.
+// returns true if there is a zip file contaning at least one entry in it.
+// returns false otherwise.
+bool containsBundleZipFile(const std::string bundle_file_path)
+{
+  try
+  {
+    ZipFile bundle(bundle_file_path);
+    return (bundle.size() > 0);
+  }
+  catch (...)
+  {
+    return false;
+  }
+  return false;
+}
+
+// Test the multiple ways in which a manifest can be added...
+// --manifest-add to a user defined zip file (using --out-file)
+// --manifest-add to a shared library (using --bundle-file)
+// --res-add to a user defined zip file (using --out-file)
+// --res-add to a shared library (using --bundle-file)
+// --zip-add to merge with another zip file (using --out-file)
+// --zip-add to append to a bundle (using --bundle-file)
+//
+// in all cases, the invalid manifest.json should never be added
+// to the zip or shared library.
+
+void testManifestAddWithInvalidJSON(const std::string& rcbinpath, const std::string& tempdir)
+{
+  std::string invalidSyntax(R"(
+		{
+		 "bundle.name": "foo",
+		 "bundle.version" : "1.0.2"
+		 "bundle.description" : "This bundle is broken!",
+		 "authors" : ["John Doe", "Douglas Reynolds", "Daniel Cannady"],
+		 "rating" : 5
+		} 
+    )");
+
+  US_TEST_NO_EXCEPTION_REQUIRED(createManifestFile(tempdir, invalidSyntax));
+
+  std::ostringstream cmd;
+  cmd << rcbinpath;
+  cmd << " --bundle-name " << "mybundle";
+  cmd << " --out-file " << tempdir << "invalid_syntax.zip";
+  cmd << " --manifest-add " << tempdir << "manifest.json";
+  US_TEST_CONDITION(BUNDLE_MANIFEST_VALIDATION_ERROR_CODE == runExecutable(cmd.str()), "Fail to embed manifest containing JSON syntax errors.");
+
+  ZipFile invalid_syntax_zip(tempdir + "invalid_syntax.zip");
+  US_TEST_CONDITION(0 == invalid_syntax_zip.size(), "Test that the invalid manifest.json file was not added");
+
+  std::string invalid_syntax_bundle_name("invalid_syntax.bundle");
+  US_TEST_NO_EXCEPTION_REQUIRED(createDummyBundle(tempdir, invalid_syntax_bundle_name));
+
+  std::ostringstream cmd2;
```

Comment:
I suggest creating a function that constructs the full command string and call it from multiple places instead of constructing them with the stream operators over and over.

Reply:
I'm not sure how useful this would be. There is nothing special about how the command line string is constructed. I'd like to defer this for now.